### DISCUSSION TITLE
Ubuntu/Debian compatibility ?

### DISCUSSION THREAD

**Discussion by Enissay** [gene-git/wg_tool#7] [General]
Before I start messing with my system installing all dependencies and such, is the script compatible with both Ubuntu/Debian ?

I have been looking for a decent management script and google lead me here <3

**Comment by gene-git** [comment]
Hi - should be completely fine - it's pure python -  ( python versions 3.9 or newer)  - so as long your python satisfies that then it should be completely fine. 

The dependencies are listed in README.   You can test it by cloning the repo and running from there. Easy way to try it out. 

Aside from some convenience python modules, it uses wireguard tools and nftables - both are needed to run vpn anyway  (or iptables if thats your thing) with or without wg-tool.

Feel free to tag me if I can help.

**Comment by Enissay** [comment]
Thank you for the fast reply.

I installed the dependencies using `pip`, now I am confused how to set things up... From the [readme](https://github.com/gene-git/wg_tool#installation), I am not sure where to run this (already cloned the project, it has no `dist` folder):
```
    rm -f dist/*
    poetry build --format wheel
    ./scripts/do-install /
```

**Comment by gene-git** [comment]
testing: 
   You can run out of the cloned repo for testing by running wg-tool.py  - try ./wg-tool.py -h

build and installing:
1) poetry build creates a python wheel (which holds all the code) and saves it in ./dist/xxx.whl
the install script would then install to destination directory prefix (/ above) - the code would then be installed in something like  /usr/lib/python3.10/site-packages/wg_tool .
 a sym link would also be installed  in /usr/bin/wg-tool pointing to the installed module wg-tool.py.

once its installed and available in /usr/bin then it would be in your path.

let me know how it goes.

**Comment by gene-git** [comment]
to keep the repo clean probably better to create a test directory and run in there.
mkdir ../test 
cd ../test
ln -s ../wg_tool/wg-tool.py wg-tool
./wg-tool -h

**Comment by gene-git** [comment]
be nice if someone made a .deb package for wg_tool

**Comment by Enissay** [comment]
A docker image would be better given all the dependencies I think...

**Comment by gene-git** [comment]
good idea on docker  ... i forgot to answer your question fully- you run poetry build in the top level of repo - and ditto for installer script which takes one arg which is the destination prefix. Those to things (build/install) are normally done by the package builder tool.

packages take care of the dependencies - so installing a package makes it simple too.

**Comment by Enissay** [comment]
I broke my system as I feared... I only had python 3.8 and meant to add 3.10... There was some conflict which was suggested to run `apt purge python3` (dumb me)...

I've been trying to fix it for the last hour in vain... I will try again tomorrow as I am useless after midnight... A docker package is certainly the way to go xD

**Comment by gene-git** [comment]
geez so sorry - thats really annoying - :(

**Comment by gene-git** [comment]
hope you can things back to normal

**Comment by gene-git** [comment]
the only need for 3.9 is using match() instead of if else ... there are a few places but it would be possible to make a version without match() statements I would think.

**Comment by gene-git** [comment]
again - so sorry

**Comment by Enissay** [comment]
Really, no worries... It probably will be less hassle installing the system from scratch (it's a remote server), it will just need some extra configs (FW, SSH, some other custom packages, ...)... 

I just hope it is worth it :P as I am still in the "testing phase" xD

**Comment by gene-git** [comment]
remote machine ... ug .. thats a lot of work.

**Comment by gene-git** [comment]
would be doable to make copies of the 2 files using match() and change them to use if else - those quite likely would work okay with python 3.8.

**Comment by gene-git** [comment]
I added 2 files in lib38 which use if/else instead of match(). These should work on python 3.8. Simply copy them over the regular ones. Hope this helps.

**Comment by Enissay** [comment]
Some feedback:
+ I started from a fresh system Ubuntu 22.04 with python 3 preinstalled `Python 3.10.6`
+ `sudo apt install wireguard` will install tools as well
+ `pip install tomli_w netaddr qrcode tomli build installer wheel poetry` => adding a requirements file would've been useful maybe ?
+ `export PATH="/home/ubuntu/.local/bin:$PATH"` make sure installed packages are in path => warning will show correct path to add if not
+ `git clone https://github.com/gene-git/wg_tool.git`
+ `cd wg_tool`
+ `poetry build --format wheel`
+ `./scripts/do-install /`
+ It fails if not ran using sudo... further in the script it spits `WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv`. Not sure what is the best path here...

```bash
$ ./scripts/do-install /
rsync: [Receiver] mkdir "//usr/share/wg_tool" failed: Permission denied (13)
rsync error: error in file IO (code 11) at main.c(787) [Receiver=3.2.3]
rsync: [Receiver] mkdir "///usr/share/licenses/wg_tool/LICENSE" failed: Permission denied (13)
rsync error: error in file IO (code 11) at main.c(787) [Receiver=3.2.3]
chmod: changing permissions of '///etc/wireguard/': Operation not permitted
rsync: [Receiver] ERROR: cannot stat destination "////etc/wireguard//scripts/": Permission denied (13)
rsync error: errors selecting input/output files, dirs (code 3) at main.c(768) [Receiver=3.2.3]

```
+ Got an error as `python` not found, so updated manually to `python3`
+ Further as well I got `pip no such option: --root-user-action` which I had removed to finish the install
+ Then the last step:
```bash
$ cd src/wg_tool/
$ mkdir ../test
$ cd ../test
$ ln -s ../wg_tool/wg-tool.py wg-tool
$ ./wg-tool -h
-bash: ./wg-tool: /usr/bin/python: bad interpreter: No such file or directory
$ vim ../wg_tool/wg-tool.py    # update to python3
$ ./wg-tool -h
usage: wg-tool [-h] [-wkd WORK_DIR] [-i] [-add] [-mod] [-dnsrch] [-dnslin] [-int] [-uuk] [-usk] [-all] [-act] [-inact] [-imp IMPORT_USER] [-keep KEEP_HIST] [-keep_wg KEEP_HIST_WG] [-l] [-rpt SHOW_RPT]
               [-rrpt] [-det] [-sop] [-v]
               [users ...]

WgTool : Manage wireguard Server & User/Profiles

positional arguments:
  users                 user_1[:prof1,prof2,...] user_2[:prof_1,prof_2]

options:
  -h, --help            show this help message and exit
  -wkd WORK_DIR, --work_dir WORK_DIR
                        Set the working dir. If unset, use dir path : /etc/wireguard/wg-tool:./
  -i, --init            Initialize: create server config - please edit. Can also set --work_dir
  -add, --add_users     Add user(s) and/or user profiles user:prof1,prof2,...
  -mod, --mod_users     Modify user(s) profiles (with -dnsrch, -dnslin)
  -dnsrch, --dns_search
                        Add/Mod user with dns search list from server config DNS_SEARCH
  -dnslin, --dns_linux  Linux. profile resolv.conf managed using PostUp/Down scripts
  -int, --int_serv      With --add_users uses internal wireguard server
  -uuk, --upd_user_keys
                        Update existing user(s) keys.
  -usk, --upd_serv_keys
                        Update server keys - affects all users
  -all, --all_users     Some opts (e.g. upd_user_keys) may apply to all users/profiles
  -act, --active        Mark users/profiles user[:profile,...] active
  -inact, --inactive    Mark users/profiles user[:profile,...] inactive
  -imp IMPORT_USER, --import_user IMPORT_USER
                        Import a user profile : --imp user.conf user_name:profile_name
  -keep KEEP_HIST, --keep_hist KEEP_HIST
                        Keep config history
  -keep_wg KEEP_HIST_WG, --keep_hist_wg KEEP_HIST_WG
                        Keep wg-config history
  -l, --list_users      List users/profiles
  -rpt SHOW_RPT, --show_rpt SHOW_RPT
                        Output of "wg show" (file, "stdin") -> connected users report
  -rrpt, --run_show_rpt
                        Run "wg show" -> connected users report (see also -rpt)
  -det, --details       Adds details to rpt and list output
  -sop, --save_opts     Save defaults for keep_hist/keep_hist_wg
  -v, --verb            Be more verbose
```

+ Finally, I went through the readme and updated the needed values. Then tried to add a test user and got the below error:
```bash
$ ./wg-tool --add_user enissay:pixel
Normal Mode: using existing setup
Adding enissay:pixel
    config:   updating - server
    config:   updating - enissay
 wg-config:   updating - enissay:pixel
Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python3.10/site-packages/qrcode/image/pil.py", line 5, in <module>
    from PIL import Image, ImageDraw
ModuleNotFoundError: No module named 'PIL'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/wg_tool/src/test/./wg-tool", line 26, in <module>
    main()
  File "/home/ubuntu/wg_tool/src/test/./wg-tool", line 22, in main
    wgtool.doit()
  File "/home/ubuntu/wg_tool/src/wg_tool/lib/class_wg.py", line 385, in doit
    write_wg_users(self)
  File "/home/ubuntu/wg_tool/src/wg_tool/lib/wg_users.py", line 131, in write_wg_users
    okay = text_to_qr_file(prof_str, actual_file,)
  File "/home/ubuntu/wg_tool/src/wg_tool/lib/utils.py", line 23, in text_to_qr_file
    img = qr.make_image(fill_color="black", back_color="white")
  File "/home/ubuntu/.local/lib/python3.10/site-packages/qrcode/main.py", line 296, in make_image
    from qrcode.image.pil import PilImage
  File "/home/ubuntu/.local/lib/python3.10/site-packages/qrcode/image/pil.py", line 7, in <module>
    import Image
ModuleNotFoundError: No module named 'Image'
```

**Comment by gene-git** [comment]
Several things clearly including needing root to install in system area.
When the python qrcode module was installed, it should have installed its own dependencies - including pillow - which is misssing - can you please install pillow and maybe let whoever packaged qrcode that it missed the pillow dep.

**Comment by gene-git** [comment]
I did assume that python points to python3 since python2 is long ago dead/deprecated.

**Comment by gene-git** [comment]
The install script should work as non-root if you do something like:
mkdir  ~/tools
./scripts/install ~/tools
then it will install everything under ~/tools and you should be able to run from there
(other issues aside for a moment)

**Comment by gene-git** [comment]
my todo
  - change python to python3
  -  remove pip option from install script
  -  add requirements  file
  -

**Comment by Enissay** [comment]
` can you please install pillow and maybe let whoever packaged qrcode that it missed the pillow dep.`

Pillow solved it.

Now getting his:
```bash
$ sudo wg-quick up wg-configs/server/wg0.conf
[#] ip link add wg0 type wireguard
[#] wg setconf wg0 /dev/fd/63
[#] ip -4 address add XXXX.1/24 dev wg0
[#] ip -6 address add YYYY::1/64 dev wg0
[#] ip link set mtu 8920 up dev wg0
[#] /usr/bin/nft -f /etc/wireguard/scripts/postup.nft
/usr/bin/wg-quick: line 295: /usr/bin/nft: No such file or directory
[#] ip link delete dev wg0
```

I will have to install nft and try again, but I am having internet issues... Will try again tomorrow...

**Comment by gene-git** [comment]
nft is nftables runtime -

**Comment by Enissay** [comment]
```
$ whereis nft
nft: /usr/sbin/nft /usr/share/man/man8/nft.8.gz
```

**Comment by gene-git** [comment]
Ah gotcha - - we need to change the script  to just 'nft' (many distros have /sbin sym link to bin) -

**Comment by gene-git** [comment]
for now can you simply edit the wg0.conf and change /usr/bin/nft to /usr/sbin/nft

I can change it to just 'nft' - that should work everywhere

**Comment by Enissay** [comment]
Yes, that's what I did.


Sadly I have no traffic on phone through the vpn tunnel... I will look further tomorrow... I am of no use past 3am xD
```
$ sudo wg show
interface: wg0
  public key: E25nseQ6GEg2oyyyyyyyyyyywSKz8HHPJEx4=
  private key: (hidden)
  listening port: 60xxx

peer: JwOgjMki5c4ljeq6vxxxxxxxxxxxxxxxxxxxIulUeZGhM=
  preshared key: (hidden)
  endpoint: XXXX.154:33338
  allowed ips: YYYY.3/32
  latest handshake: 12 seconds ago
  transfer: 2.54 KiB received, 124 B sent
```

**Comment by gene-git** [comment]
The correct fix is to change your server config  

config/server/server.conf

and change the /usr/bin/nft to /usr/sbin/nft
that will fix it permanently for you.

Wow - tis late there!!!

**Comment by gene-git** [comment]
also please try instead of wg show:
sudo wg-tool -rrpt

**Comment by gene-git** [comment]
If no traffic then (assuming vpn server itself has internet / local access) - the nftables config is likely not right for your setup

**Comment by gene-git** [comment]
I will push the python3 change and the pip option fix to gh repo -

**Comment by gene-git** [comment]
Can you check the postup script to be sure you have the right network settings - the default script should work as long as your vpn server has internet access - but it does need to be edited for your own network,.

thanks

**Comment by gene-git** [comment]
Still on my todo list - create requirements file.  

Please Let me know if you got the postup script working.  Thanks!

**Comment by gene-git** [comment]
requirements.txt added in commit [7a548d2](https://github.com/gene-git/wg_tool/commit/7a548d2a6736dba8544385572cb99f3f068e987d)

**Comment by Enissay** [comment]
> Please Let me know if you got the postup script working. Thanks!

Sadly not. Sorry for my disappearance but I am on holidays (since over a week actually)... I'll be back home later this week and will dig deeper to find a fix

**Comment by gene-git** [comment]
Sorry to hear that - happy to help where I can. I assume vpn server has internet access so when the vpn clients get nat'ed their traffic should get to internet as well (and would appear to come from vpn server of course). Have wonderful rest of holidays and very merry christmas !

**Comment by gene-git** [comment]
May be worth double checking that the  network settings look right in your postup nftables script.

**Comment by Enissay** [comment]
No worries, last time I checked, server had internet and handshake was successful, yet no internet on the client... 
I admit some of the options I just winged (I was not sure)... When back, I will try again from scratch which will allow me also to try all the changes you generously implemented... I will get back with questions to set things right...

**Comment by gene-git** [comment]
ok that sounds good  - have fun day :)

**Comment by gene-git** [comment]
(it really does sound like the postup script is worth looking over)

**Comment by gene-git** [comment]
one more thought occurred to me - please make sure ip forwarding is turned on 
sysctl -w net.ipv4.ip_forward=1
and add line to /etc/sysctl.d/systctl.conf (or whatever name you prefer for file) :
net.ipv4.ip_forward = 1

**Comment by gene-git** [comment]
Hi any updates from your side?

**Reply by Enissay** [reply]
Oh, sadly not. I got distracted since with other tasks :<
I am planning however to dive in again this month still. I will revisit and update..

**Reply by gene-git** [reply]
Understand  - life can certainly keep you busy, especially with the unexpected!
Happy to help if I can.
